### PR TITLE
Sparse Baselines for Transformers

### PR DESCRIPTION
**Pull request by mvacaporale**
This PR include results for the static sparse baselines:
* sparse_v2_100k - static sparsity through all linear layers in the encoder
* sparse_v3_100k - like v2, but the token embeddings are sparse as well

Corresponding to v3, there's a new `SparseEmbeddings` module and a `FullyStaticSparseBertModel` class that uses this module to sparsify its embeddings.

### CODE REVIEW

**Comment by lucasosouza** [comment]
@subutai the results for static sparse baseline are in this JIRA, I pasted below

|            | average_bert     | average_glue      | cola           | mnli                         | mrpc        | qnli     | qqp         | rte      | sst2     | stsb                  | wnli     | perplexity |
|:-----------|:-----------------|:------------------|:---------------|:-----------------------------|:------------|:---------|:------------|:---------|:---------|:----------------------|:---------|:-----------|
|            | Average w/o wnli | Average all tasks | Matthew's corr | Matched acc./Mismatched acc. | F1/Accuracy | Accuracy | Accuracy/F1 | Accuracy | Accuracy | Person/Spearman corr. | Accuracy |            |
| bert_1mi |          80.76 |          76.82 |  48.87 | 84.08/84.57 | 89.76/85.68 |  91.19 | 90.58/87.17 | 66.02 |  91.44 | 87.67/87.54 |  45.31 | 5.013 |
| bert_100k |          75.71 |          72.51 |  40.98 | 78.26/78.65 | 84.05/77.86 |  87.74 | 89.12/85.25 |  58.2 |  88.31 | 83.90/83.80 |  46.88 | 8.619 |
| sparse_v1_100k |          72.67 |          67.91 |  22.43 | 77.25/77.96 | 85.37/78.82 |  84.65 | 88.23/84.32 | 58.32 |  87.27 | 82.79/82.65 |  29.84 | 10.461 |
| sparse_v2_100k |          68.41 |          67.06 |  15.96 | 72.25/73.11 | 80.50/71.35 |  81.23 | 85.78/81.01 | 58.98 |  82.64 | 76.53/76.40 |  56.25 | 16.696 |
| sparse_v3_100k |          68.25 |          66.92 |  16.35 | 72.00/72.32 | 80.42/70.83 |  80.31 | 85.04/80.49 | 58.98 |  82.52 | 76.96/77.60 |  56.25 | 22.81 |

<br/><br/>

**Legend**
* *sparse_v1:* Static sparse encoder with only the non-attention linear layers sparsified (`model_type=static_sparse_non_attention_bert`).~41% sparse
* *sparse_v2:* Static sparse encoder with all linear layers sparsified, including attention (`model_type=static_sparse_encoder_bert`) ~63% sparse
* *sparse_v3:* Static sparse encoder (as with v2) as well as static sparse token embeddings (`model_type=fully_static_sparse_bert`) ~79% sparse

File: projects/transformers/README.md [hunk 1]
```diff
@@ -13,13 +13,15 @@ In progress, current results:
 | bert_700k |          79.77 |          75.25 |  47.73 | 83.91/84.22 | 88.27/83.59 |  91.51 | 90.37/86.90 | 62.11 |  91.44 | 86.88/86.70 |  39.06 |  |
 | bert_100k |          75.71 |          72.51 |  40.98 | 78.26/78.65 | 84.05/77.86 |  87.74 | 89.12/85.25 |  58.2 |  88.31 | 83.90/83.80 |  46.88 | 8.619 |
 | sparse_v1_100k |          72.67 |          67.91 |  22.43 | 77.25/77.96 | 85.37/78.82 |  84.65 | 88.23/84.32 | 58.32 |  87.27 | 82.79/82.65 |  29.84 | 10.461 |
-| sparse_v2_100k |  |  |  |  |  |  |  |  |  |  |  |  |
+| sparse_v2_100k |          68.41 |          67.06 |  15.96 | 72.25/73.11 | 80.50/71.35 |  81.23 | 85.78/81.01 | 58.98 |  82.64 | 76.53/76.40 |  56.25 | 16.696 |
+| sparse_v3_100k |          68.25 |          66.92 |  16.35 | 72.00/72.32 | 80.42/70.83 |  80.31 | 85.04/80.49 | 58.98 |  82.52 | 76.96/77.60 |  56.25 | 22.81 |
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/transformers/README.md:17]: It is also strange wnli results are so high in v2 and v3 if we haven't made any changes to the finetuning regime - what could cause that sudden increase?

File: projects/transformers/models/static_sparse_bert.py [hunk 2]
```diff
@@ -136,3 +139,110 @@ def sparsify_model(self):
...
+
+    def __init__(self, config, add_pooling_layer=True):
+        # Call the init one parent class up. Otherwise, the model will be defined twice.
+        BertPreTrainedModel.__init__(self, config)
+        self.config = config
+
+        self.embeddings = BertEmbeddings(config)
+        self.encoder = BertEncoder(config)
+
+        self.pooler = BertPooler(config) if add_pooling_layer else None
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply projects/transformers/models/static_sparse_bert.py:166]: In masked language modeling, the pooler is None. They call `self.bert = BertModel(config, add_pooling_layer=False)` in `BertForMaskedLM`. It is however used in sequence classification. In the `BertPooler` class, it says "We "pool" the model by simply taking the hidden state corresponding to the first token." I presume this output is used to classify the sequence appropriately. We shouldn't need to sparsify this correct?

[the remark this answers is not in the document]
Comment (mvacaporale) [reply projects/transformers/models/static_sparse_bert.py:166]: Per your second point: We could make a network that makes that distinction. It wouldn't be too hard to make a new BertModel or modify an existing one when needed.

File: projects/transformers/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mvacaporale) [reply projects/transformers/README.md:17]: Could the high WNLI accuracy just be variance?

[the remark this answers is not in the document]
Comment (mvacaporale) [reply projects/transformers/README.md:17]: I'm gonna run WNLI again and see if I get something high again.

[the remark this answers is not in the document]
Comment (mvacaporale) [reply projects/transformers/README.md:17]: Per the perplexity comment: It's possible that since perplexity involves an exponent (as in $2^{H(P, Q)}), the difference between values could be more exaggerated as it increases. So maybe a jump in perplexity from 8 to 10 is roughly the same as a jump from 16 to 22. That it is to say, it's hard to compare perplexities at face value if they are not on a linear scale. My understanding of how they are calculated is still limited, so maybe I'm wrong on this. I'll try and think more about how to interpret their values.

[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/transformers/README.md:17]: Good point on the perplexity not being on linear scale. 

> I'm gonna run WNLI again and see if I get something high again.

No need to run again. Remember all values are saved, and the reduction is only performed at the export_results script. You can inspect by loading the pickled class and printing all results, and that should tell whether it is a matter of variance or not. If that is the case, the means should be more coherent than the maxes.

File: projects/transformers/experiments/sparse_bert.py [hunk 3]
```diff
@@ -54,6 +54,19 @@
     trainer_callbacks=[RezeroWeightsCallback()]
 )
 
+
+fully_static_sparse_bert_100k = deepcopy(bert_100k)
+fully_static_sparse_bert_100k.update(
+    # Model Arguments
+    overwrite_output_dir=False,
+    model_type="fully_static_sparse_bert",
+    config_kwargs=dict(
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/transformers/experiments/sparse_bert.py:63]: Good point. Probably so.

File: projects/transformers/models/static_sparse_bert.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/transformers/models/static_sparse_bert.py:166]: I actually meant the last layer in cls, my confusion in calling it the pooler. That layer is not discarded and being considered in the number of params of the model, right? If that is the case, shouldn't we consider sparsifying as well? 

image

File: projects/transformers/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/transformers/README.md:17]: You can even modify the TaskResults to have new reduction methods, including for example mean with margin of error. The original intention in making it a class was to add these new functionalities as required.

[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/transformers/README.md:17]: Although I recognize it would be cleaner to save it as raw human readable data and have reductions methods in a separate module that handles the data. (todo)

[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/transformers/README.md:17]: On the perplexity being on an exponential scale - if it fits our intuition better, we can track log(ppl) along with ppl.

File: projects/transformers/models/static_sparse_bert.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mvacaporale) [reply projects/transformers/models/static_sparse_bert.py:166]: The `cls` network also seems to be discarded during sequence classification. The one constant throughout all the scenarios is Bert itself belonging to `model.bert`. I think this is what we'll solely want to sparsify. Do you agree?

[the remark this answers is not in the document]
Comment (mvacaporale) [reply projects/transformers/models/static_sparse_bert.py:166]: Note: There is about 0.2% more sparsity in the `BertModel` that can come from the `position_embeddings` and `token_embeddings`. I haven't added these just yet.

File: projects/transformers/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mvacaporale) [reply projects/transformers/README.md:17]: I think we already do. It's the loss I believe. From our `run_utils.py` code:
```
perplexity = math.exp(eval_output["eval_loss"])
```

[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/transformers/README.md:17]: That is strange. We are using regular cross entropy (negative loglikelihood). That means perplexity is the exponential of the cross entropy loss? It doesn't match my understanding - we should definitely look into this at some point. 

Also, not clear to me how perplexity is calculated for masked language model, it is not a very direct application. We talked about this when we were discussing perplexity.

File: projects/transformers/models/static_sparse_bert.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/transformers/models/static_sparse_bert.py:166]: If you are positive that specific layer is being discarded (the intermediate layer of cls), then I agree. But we came to a different conclusion in our discussions on the model and number of parameters about two weeks ago.

[the remark this answers is not in the document]
Comment (mvacaporale) [reply projects/transformers/models/static_sparse_bert.py:166]: As far as I can tell, only the `BertForSequenceClassification` model is being used for fine-tuning tasks. This is defined as
```
class BertForSequenceClassification(BertPreTrainedModel):
    def __init__(self, config):
        super().__init__(config)
        self.num_labels = config.num_labels

        self.bert = BertModel(config)
        self.dropout = nn.Dropout(config.hidden_dropout_prob)
        self.classifier = nn.Linear(config.hidden_size, config.num_labels)

        self.init_weights()
```

This shows there is no `self.cls`, so no weights are being loaded from `cls` used in masked lm.

[the remark this answers is not in the document]
Comment (mvacaporale) [reply projects/transformers/models/static_sparse_bert.py:166]: One possible source of discrepancy is the case vs uncased. If we were expecting Bert-case to have more parameters, which in fact it doesn't, then we'd be inclined to include other parameters from other parts of the masked-lm network to get to the expected sum.
